go语言智能合约，调用NewIterator报错

**Brief of the issue**
go语言的NewIterator接口，得到的iterator之后调用Next()方法，超级链报错，并直接把超级链跑挂
报错信息
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0xf83157]

goroutine 267 [running]:
github.com/xuperchain/xuperunion/xvm/exec.CaptureTrap(0xc013f35410)
	/root/xuperunion/xvm/exec/trap.go:64 +0x113
panic(0x1602740, 0x2528360)
	/opt/go/src/runtime/panic.go:522 +0x1b5
github.com/xuperchain/xuperunion/xmodel.(*XModel).Select(0x0, 0xc013f7aba4, 0xc, 0xc013f7bae0, 0xb, 0x10, 0xc013f7baf0, 0xb, 0x10, 0x4, ...)
	/root/xuperunion/xmodel/xmodel.go:228 +0x187
github.com/xuperchain/xuperunion/xmodel.(*XMCache).NewXModelCacheIterator(0xc013f7fbc0, 0xc013f7aba4, 0xc, 0xc013f7bae0, 0xb, 0x10, 0xc013f7baf0, 0xb, 0x10, 0x1a79ca0, ...)
	/root/xuperunion/xmodel/xmodel_cache_iterator.go:56 +0x478
github.com/xuperchain/xuperunion/xmodel.(*XMCache).Select(...)
	/root/xuperunion/xmodel/xmodel_cache.go:198
github.com/xuperchain/xuperunion/contract/bridge.(*SyscallService).NewIterator(0xc00012d820, 0x1a78ee0, 0xc0000360d0, 0xc013f8b200, 0x0, 0x0, 0x0)
	/root/xuperunion/contract/bridge/syscall_service.go:264 +0xf0
reflect.Value.call(0xc000359500, 0xc00012a660, 0x13, 0x1775c34, 0x4, 0xc013f34e98, 0x3, 0x3, 0x1702d60, 0xc013f84501, ...)
	/opt/go/src/reflect/value.go:447 +0x461
reflect.Value.Call(0xc000359500, 0xc00012a660, 0x13, 0xc013f34e98, 0x3, 0x3, 0x0, 0x203004, 0x2)
	/opt/go/src/reflect/value.go:308 +0xa4
github.com/xuperchain/xuperunion/contract/bridge/memrpc.(*Server).CallMethod(0xc013f230e0, 0x1a78ee0, 0xc0000360d0, 0x2, 0xc013f34f90, 0xb, 0x7f4c43bff040, 0x1e, 0x33f9dfc0, 0x10bbdc0, ...)
	/root/xuperunion/contract/bridge/memrpc/server.go:98 +0x42d
github.com/xuperchain/xuperunion/contract/wasm/vm/xvm.(*syscallResolver).goCallMethod(0xc013ed1420, 0x1a842c0, 0xc013f56cc0, 0xc06db08, 0xc000000000)
	/root/xuperunion/contract/wasm/vm/xvm/resolver.go:63 +0x1f9
github.com/xuperchain/xuperunion/xvm/exec.applyFuncCall(0x1a842c0, 0xc013f56cc0, 0x15d2ec0, 0xc013ed1440, 0xc013f7bad8, 0x1, 0x1, 0x100000000)
	/root/xuperunion/xvm/exec/resolver.go:153 +0xb8
github.com/xuperchain/xuperunion/xvm/exec.xvm_call_func(0x1, 0x15, 0xc013f56cc0, 0x7f4c9fffe920, 0xc000000001, 0x30)
	/root/xuperunion/xvm/exec/resolver.go:131 +0x13d
github.com/xuperchain/xuperunion/xvm/exec._cgoexpwrap_de3f9b986089_xvm_call_func(0x1, 0x15, 0xc013f56cc0, 0x7f4c9fffe920, 0x1, 0x7f4c90001900)
	_cgo_gotypes.go:371 +0x51
github.com/xuperchain/xuperunion/xvm/exec._Cfunc_xvm_call(0xc013f56cc0, 0x7f4c90001640, 0xc013f7b380, 0x2, 0xc013f7b398, 0xc000000000)
	_cgo_gotypes.go:227 +0x4d
github.com/xuperchain/xuperunion/xvm/exec.(*aotContext).Exec.func2(0xc013f56cc0, 0x7f4c90001640, 0xc013f7b380, 0xc013f7b380, 0x2, 0x2, 0xc013f7b398, 0x20)
	/root/xuperunion/xvm/exec/context.go:129 +0xcc
github.com/xuperchain/xuperunion/xvm/exec.(*aotContext).Exec(0xc013f56cc0, 0x177571d, 0x3, 0xc013f7b380, 0x2, 0x2, 0x0, 0x0, 0x0)
	/root/xuperunion/xvm/exec/context.go:129 +0x185
github.com/xuperchain/xuperunion/contract/wasm/vm/xvm.(*xvmInstance).Exec(0xc013f84480, 0x177571d, 0x3, 0xc013f35498, 0x97a088)
	/root/xuperunion/contract/wasm/vm/xvm/instance.go:61 +0xa3
github.com/xuperchain/xuperunion/contract/wasm.(*bridgeInstance).Exec(0xc013f8d020, 0xf9932b, 0x1689d00)
	/root/xuperunion/contract/wasm/bridge_instance.go:38 +0x9e
github.com/xuperchain/xuperunion/contract/bridge.(*vmContextImpl).Invoke(0xc013f8d040, 0xc013f7abb0, 0x5, 0xc013f7ecc0, 0x0, 0x0, 0x0)
	/root/xuperunion/contract/bridge/adaptor.go:40 +0x91
github.com/xuperchain/xuperunion/utxo.(*UtxoVM).verifyTxRWSets(0xc0002bb200, 0xc013db0d80, 0xc013f7eff0, 0x1, 0x0)
	/root/xuperunion/utxo/tx_verification.go:557 +0x108a
github.com/xuperchain/xuperunion/utxo.(*UtxoVM).ImmediateVerifyTx(0xc0002bb200, 0xc013db0d80, 0x9e2800, 0x59af10, 0xc013f96000, 0xc003fd1908)
	/root/xuperunion/utxo/tx_verification.go:122 +0xe5c
github.com/xuperchain/xuperunion/utxo.(*UtxoVM).VerifyTx(0xc0002bb200, 0xc013db0d80, 0x20, 0x1585540, 0x2586e41)
	/root/xuperunion/utxo/utxo.go:1318 +0x87
```

**Your environment**
master-d4f2e09 2019-12-17-16:23:36

--- a/contract/bridge/syscall_service.go
+++ b/contract/bridge/syscall_service.go
@@ -273,7 +273,7 @@ func (c *SyscallService) NewIterator(ctx context.Context, in *pb.IteratorRequest
 	out := new(pb.IteratorResponse)
 	for iter.Next() && limit > 0 {
 		out.Items = append(out.Items, &pb.IteratorItem{
-			Key:   append([]byte(""), iter.Key()...), //make a copy
+			Key:   append([]byte(""), iter.Data().GetPureData().GetKey()...), //make a copy
 			Value: append([]byte(""), iter.Data().GetPureData().GetValue()...),
 		})
 		limit -= 1

--- a/contractsdk/go/exec/kv_iterator.go
+++ b/contractsdk/go/exec/kv_iterator.go
@@ -17,6 +17,7 @@ type kvIterator struct {
 	curBuf       *pb.IteratorItem   // pointer of current position
 	curIdx       int                // next index
 	c            *contractContext   // where we can get the kv items
+	end          bool
 	err          error
 	start, limit []byte
 }
@@ -45,13 +46,19 @@ func (ki *kvIterator) load() {
 		ki.err = err
 		return
 	}
+
 	if len(resp.Items) == 0 {
-		ki.start = ki.limit
+		ki.end = true
 		return
 	}
+	lastKey := resp.Items[len(resp.Items)-1].Key
+	newStartKey := make([]byte, len(lastKey)+1)
+	copy(newStartKey, lastKey)
+	newStartKey[len(lastKey)] = 1
+
 	ki.curIdx = 0
 	ki.buf = resp.Items
-	ki.start = resp.Items[len(resp.Items)-1].Key
+	ki.start = newStartKey
 }
 
 func (ki *kvIterator) Key() []byte {
@@ -63,17 +70,22 @@ func (ki *kvIterator) Value() []byte {
 }
 
 func (ki *kvIterator) Next() bool {
-	//永远保证有数据
-	if ki.curIdx == len(ki.buf) {
-		ki.load()
-	}
-	if len(ki.buf) == 0 || ki.err != nil {
+	if ki.end || ki.err != nil {
 		return false
 	}
+	//永远保证有数据
+	if ki.curIdx >= len(ki.buf) {
+		ki.load()
+		if ki.end || ki.err != nil {
+			return false
+		}
+	}
+
 	ki.curBuf = ki.buf[ki.curIdx]
 	ki.curIdx += 1
 	return true
 }
+
 func (ki *kvIterator) Error() error {
 	return ki.err
 }

--- a/xmodel/xmodel_cache_iterator.go
+++ b/xmodel/xmodel_cache_iterator.go
@@ -53,9 +53,13 @@ func (mc *XMCache) NewXModelCacheIterator(bucket string, startKey []byte, endKey
 	iters = append(iters, mcoi)
 	mcii := mc.inputsCache.NewIterator(&util.Range{Start: rawStartKey, Limit: rawEndKey})
 	iters = append(iters, mcii)
-	mi, err := mc.model.Select(bucket, startKey, endKey)
-	if err != nil {
-		return nil, err
+	var mi Iterator
+	if mc.isPenetrate {
+		var err error
+		mi, err = mc.model.Select(bucket, startKey, endKey)
+		if err != nil {
+			return nil, err
+		}
 	}
 	return &XMCacheIterator{
 		mIter:     mi,
@@ -179,7 +183,9 @@ func (mci *XMCacheIterator) Release() {
 		return
 	}
 	mci.dir = dirReleased
-	mci.mIter.Release()
+	if mci.mIter != nil {
+		mci.mIter.Release()
+	}
 	for _, it := range mci.iters {
 		it.Release()
 	}

